install: Try to use ASDF_DATA_DIR

mook-as opened the pull request:
This allows the GPG keyrings to be installed into `${ASDF_DATA_DIR}` rather than `${ASDF_DIR}`; this is useful if asdf itself may be replaced wholesale (e.g. by a package manager).  Semantically, it also makes more sense to put the keyring into the data directory, as it is not part of the core asdf application.

Fixes #205.

augustobmoura commented on bin/install:
```diff
@@ -237,8 +237,12 @@ download_and_verify_checksums() {
-      if [ -z "${GNUPGHOME:-}" ] && [ -d "${ASDF_DIR:-$HOME/.asdf}/keyrings/nodejs" ]; then
-        export GNUPGHOME="${ASDF_DIR:-$HOME/.asdf}/keyrings/nodejs"
+      if [ -z "${GNUPGHOME:-}" ]; then
+        if [ -d "${ASDF_DIR:-$HOME/.asdf}/keyrings/nodejs" ]; then
+          export GNUPGHOME="${ASDF_DIR:-$HOME/.asdf}/keyrings/nodejs"
+        elif [ -d "${ASDF_DATA_DIR:-$HOME/.asdf}/keyrings/nodejs" ]; then
+          export GNUPGHOME="${ASDF_DATA_DIR:-$HOME/.asdf}/keyrings/nodejs"
+        fi
```
I think it's more idiomatic to put the directories candidates into a loop so we can add more candidates later and get a more explicit priority, like so:
```suggestion
        for dir in "$ASDF_DIR" "$HOME/.asdf" "$ASDF_DATA_DIR"; do
          if [ -d "$dir/keyrings/nodejs" ]; then
            export GNUPGHOME="$dir/keyrings/nodejs"
          fi
        done
```

augustobmoura replied:
Also, I would add a ASDF_NODEJS custom variable so it can be even more flexible

```suggestion
        for dir in "$ASDF_NODEJS_KEYRING_DIR" "$ASDF_DIR" "$HOME/.asdf" "$ASDF_DATA_DIR"; do
          if [ -n "$dir" ] && [ -d "$dir/keyrings/nodejs" ]; then
            export GNUPGHOME="$dir/keyrings/nodejs"
          fi
        done
```

augustobmoura replied:
If you go with the ASDF_NODEJS_KEYRING_DIR idea, we should also add some documentation in the README explaining the functionality

mook-as commented:
Added the ability to use `ASDF_NODEJS_KEYRING_DIR` instead of the default directories. Note though:
- This does _not_ expect the `keyrings/nodejs` subdirectory, since the variable name already includes all that and it wouldn't make sense.
- This would need to be set every time you run `asdf install nodejs …`, since unlike `ASDF_{DATA_,}DIR` there is no sensible value by default.

Added a bit in the readme, hopefully that's clear enough.

augustobmoura commented:
Sorry for the delay @mook-as, I will try to get into the PR this week

Stratus3D commented:
Ping @augustobmoura. Are you going to be able to test this out?

Stratus3D reviewed (approved):
I've been out of the loop on this project, but the changes here appear good to me.

augustobmoura commented:
Sorry for the delay @Stratus3D, the notification of your comment got lost between some others

The code works. It's just a bit convoluted. I'll make a new code review the old one stalled

augustobmoura commented on bin/install:
```diff
@@ -237,8 +237,17 @@ download_and_verify_checksums() {
     (
-      if [ -z "${GNUPGHOME:-}" ] && [ -d "${ASDF_DIR:-$HOME/.asdf}/keyrings/nodejs" ]; then
-        export GNUPGHOME="${ASDF_DIR:-$HOME/.asdf}/keyrings/nodejs"
+      if [ -z "${GNUPGHOME:-}" ]; then
+        if [ -n "${ASDF_NODEJS_KEYRING_DIR:-}" ]; then
+          export GNUPGHOME="$ASDF_NODEJS_KEYRING_DIR"
+        else
+          for dir in "${ASDF_DATA_DIR:-$HOME/.asdf}" "${ASDF_DIR:-$HOME/.asdf}" ;do
+            if [ -d "$dir/keyrings/nodejs" ]; then
+              export GNUPGHOME="$dir/keyrings/nodejs"
+              break
+            fi
+          done
+        fi
```
I think we can simplify this code to be a single loop, we can also use bash brace expansions to simplify some code, how about:
```suggestion
      for dir in "$ASDF_NODEJS_KEYRING_DIR" "$GNUPGHOME" {"$ASDF_DATA_DIR","$ASDF_DIR","$HOME/.asdf}"/keyrings/nodejs"; do
        if [ -d "$dir" ]; then
          export GNUPGHOME="$dir"
          break
        fi
      done
```

augustobmoura commented on bin/import-release-team-keyring:
```diff
@@ -4,6 +4,10 @@ set -o nounset -o pipefail -o errexit
 
 source "$(dirname "$0")/../lib/utils.sh"
 
+if [ -n "${ASDF_NODEJS_KEYRING_DIR:-}" ]; then
+  export GNUPGHOME="$ASDF_NODEJS_KEYRING_DIR"
+fi
+
 ## Keys from https://github.com/nodejs/node/#release-keys
```
I think this should follow the same logic as in `bin/install`

augustobmoura commented:
As of the last PR, we are not using GPG for verifying integrity anymore, so this PR should be obsolete. Thanks for the contributions anyway @mook-as!
